File: pt5/1144.decrease-elements-to-make-array-zigzag.go
package main

func movesToMakeZigzag(nums []int) int {
	n := len(nums)

	costForValleys := func(parity int) int {
		moves := 0
		for i, value := range nums {
			if i%2 != parity {
				continue
			}
			allowed := 1 << 30
			if i > 0 && nums[i-1]-1 < allowed {
				allowed = nums[i-1] - 1
			}
			if i+1 < n && nums[i+1]-1 < allowed {
				allowed = nums[i+1] - 1
			}
			if value > allowed {
				moves += value - allowed
			}
		}
		return moves
	}

	evenCost, oddCost := costForValleys(0), costForValleys(1)
	if evenCost < oddCost {
		return evenCost
	}
	return oddCost
}

File: pt5/1144.decrease-elements-to-make-array-zigzag.py
from __future__ import annotations

from typing import List


class Solution:
    def movesToMakeZigzag(self, nums: List[int]) -> int:
        n = len(nums)

        def cost_for_valleys(parity: int) -> int:
            moves = 0
            for i, value in enumerate(nums):
                if i % 2 != parity:
                    continue
                left = nums[i - 1] if i > 0 else float("inf")
                right = nums[i + 1] if i + 1 < n else float("inf")
                allowed = min(left, right) - 1
                if value > allowed:
                    moves += value - allowed
            return moves

        return min(cost_for_valleys(0), cost_for_valleys(1))

